### PR TITLE
Ep/param validations

### PR DESCRIPTION
**Pull request by eliasjpr**
### Requirements

- Allow parameters whitelisting. 
- Allow additional and friendlier validation predicates.

### Description of the Change

- With this PR we introduce Validations for params 

### Alternate Designs

One caveat is that at the moment of defining the validation rules validations are ran for those rules immediately. It would be nice to run the rules every time `validate!` or `validate` is called.

A better design would be to have this more OOP. With **Validator**, **Rules** and **Errors** objects.

### Why Should This Be In Core?

- A malicious user can exploit this scenario by sending this extra parameter to our application, thereby making themselves an administrator.
- Validations perform before hitting the model prevents additional processing. 

### Benefits

Validation rules can be defined as following

```crystal 
# setup validation rules here
# Note this runs the validation immediately
params.validation do
  required(:email) { |p| p.url? }
end
```

Returns validated params if valid otherwise raises error
```crystal
user = User.new params.validate!
```

Validations can be checked with 
```crystal
# Usage 2/ input may be invalid
unless params.valid?
  # return an error then quit this method
  response.puts {errors: params.errors}.to_json
  response.status_code 400
  return
end
```

### CODE REVIEW

File: src/amber/validations/params.cr [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    # Setups validation rules to be performed
+    #
+    # ```crystal
+    # params.validation do
+    #   required(:email) { |p| p.url? }
+    #   required(:age, UInt32)
+    # end
+    # ```
+    # returns Validator
```
Comment (bew) [inline src/amber/validations/params.cr]: The return value should not be documented

File: src/amber/validations/params.cr [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    #   response.puts {errors: params.errors}.to_json
+    #   response.status_code 400
+    # end
+    # ```
+    # returns Boolean
+    def valid?
+      errors.empty?
+    end
+
+    # Captures and Prforms the validation rules block
```
Comment (bew) [inline src/amber/validations/params.cr]: typo `Prforms`

File: spec/amber/validations/params_spec.cr [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      it "validates required field" do
+        http_params = HTTP::Params.parse("name=elias&last_name=perez&middle=j")
+        validator = Validators::Params.new(http_params)
+
+        result = validator.validation do
+          required("name") { |v| v.str? & !v.empty? }
+          required("last_name") { |v| v.str? & !v.empty? }
+        end
+
+        result.valid?.should eq true
```
Comment (bew) [inline spec/amber/validations/params_spec.cr]: `result.valid?.should be_true`

File: src/amber/validations/params.cr [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    # Validates each field with a given set of predicates returns true if the
+    # field is valid otherwise returns false
+    #
+    # ```crystal
+    # required(:email) { |p| p.email? & p.size.between? 1..10 }
+    # ```
+    def required(key, msg : String? = nil)
+      unless raw_params.has_key?(key)
+        errors[key] = {"invalid", "Param [nonexisting] does not exist."}
```
Comment (bew) [inline src/amber/validations/params.cr]: What is `"invalid"`?

Comment (bew) [inline src/amber/validations/params.cr]: Also, the following error could be rewritten as `"Param [#{key}] is required."`

File: src/amber/support/predicates.cr [hunk 5]
```diff
@@ -0,0 +1,174 @@
+module Amber
+  module Support
+    module Predicates
+      module String
+        macro included
```
Comment (bew) [inline src/amber/support/predicates.cr]: why do you need this macro?

Comment (eliasjpr) [inline src/amber/support/predicates.cr]: It should be a module my bad

Comment (eliasjpr) [inline src/amber/support/predicates.cr]: It should be a module my bad

Comment (eliasjpr) [inline src/amber/support/predicates.cr]: Actually no, the intention of the macro included is to extend to monkey path the string class with new predicte methods

File: src/amber/validations/params.cr [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    #   response.status_code 400
+    # end
+    # ```
+    def valid?
+      errors.empty?
+    end
+
+    # Captures and performs the validation rules block
+    #
+    # returns Nil
```
Comment (bew) [inline src/amber/validations/params.cr]: Still a useless return value documentated

File: src/amber/validations/params.cr [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    # Validates each field with a given set of predicates returns true if the
+    # field is valid otherwise returns false
+    #
+    # ```crystal
+    # required(:email) { |p| p.email? & p.size.between? 1..10 }
+    # ```
+    def required(key, msg : String? = nil)
+      unless raw_params.has_key?(key)
+        errors[key] = {"", "Param [#{key}] does not exist."}
```
Comment (bew) [inline src/amber/validations/params.cr]: You removed `invalid` but what is this `""` value? what does it represent?

File: src/amber/support/predicates.cr (hunk 5 above)
Comment (eliasjpr) [inline src/amber/support/predicates.cr]: Macro included not needed here. Will be removed

File: src/amber/validations/params.cr (hunk 7 above)
Comment (eliasjpr) [inline src/amber/validations/params.cr]: this holds the value for the `param[key]` but in this case there is no key so there is no value

Comment (bew) [inline src/amber/validations/params.cr]: You should put `nil` do better differenciate things?

Comment (eliasjpr) [inline src/amber/validations/params.cr]: Will do

File: src/amber/validations/params.cr [hunk 8]
```diff
@@ -0,0 +1,103 @@
+require "../support/predicates"
+
+# We are patching the String class and Number struct to extend the predicates
+# available this will allow to add friendlier methods for validation cases.
+class String
```
Comment (fridgerator) [inline src/amber/validations/params.cr]: `support/predicates.cr` already patches these, does it need to be duplicated?

Comment (eliasjpr) [inline src/amber/validations/params.cr]: Good catch @fridgerator. I did some cleanup on this.

**Comment by eliasjpr** [comment]
Please keep in mind that I will circle back to this implementation to make it more OOP as @bew suggested. Right now this is a working proof of concept as we figure out design matters.

**Comment by bew** [comment]
Good job! :smile:

**Pull request merged**